osv/ecosystems/haskell.py

```diff
@@ -0,0 +1,155 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Haskell ecosystem helpers.
+
+Contact the Haskell Security Response Team <security-advisories@haskell.org>
+if something is broken and you need help to fix it.
+
+"""
+
+import re
+import requests
+import typing
+
+from . import config
+from .helper_base import Ecosystem, EnumerateError
+from .. import semver_index
+
+
+class Hackage(Ecosystem):
+  """Hackage (Haskell package index) ecosystem."""
+
+  _API_PACKAGE_URL = 'https://hackage.haskell.org/package/{package}.json'
+
+  def sort_key(self, version):
+    """Sort key.
+
+    The Haskell package version data type is defined at
+    https://hackage.haskell.org/package/Cabal-syntax/docs/Distribution-Types-Version.html
+
+    """
+    return [int(x) for x in version.split('.')]
+
+  def enumerate_versions(self,
+                         package,
+                         introduced,
+                         fixed=None,
+                         last_affected=None,
+                         limits=None):
+    """Enumerate versions."""
+    response = requests.get(
+        self._API_PACKAGE_URL.format(package=package), timeout=config.timeout)
+    if response.status_code == 404:
+      raise EnumerateError(f'Package {package} not found')
+    if response.status_code != 200:
+      raise RuntimeError(
+          f'Failed to get Hackage versions for {package} with: {response.text}')
+
+    response = response.json()
+    versions = list(response.keys())
+
+    self.sort_versions(versions)
+    return self._get_affected_versions(versions, introduced, fixed,
+                                       last_affected, limits)
+
+
+class GHC(Ecosystem):
+  """Glasgow Haskell Compiler (GHC) ecosystem."""
+
+  _API_PACKAGE_URL = ('https://gitlab.haskell.org'
+                      '/api/v4/projects/3561/repository/tags?per_page=-1')
+  """
+  Historical versions do not have tags in the Git repo, so we hardcode the
+  list.  See https://github.com/google/osv.dev/pull/1463 for discussion.
+  """
+  HISTORICAL_VERSIONS = [
+    '0.29',
+    '2.10',
+    '3.02', '3.03',
+    '4.02', '4.04', '4.06', '4.08', '4.08.1', '4.08.2',
+    '5.00', '5.00.1', '5.00.2', '5.02', '5.02.1', '5.02.2', '5.02.3',
+    '5.04', '5.04.1', '5.04.2', '5.04.3',
+    '6.0', '6.0.1',
+    '6.2', '6.2.1', '6.2.2',
+    '6.4', '6.4.1', '6.4.2', '6.4.3',
+    '6.6', '6.6.1',
+    '6.8.1', '6.8.1', '6.8.3',
+    '6.10.1', '6.10.2-rc1', '6.10.2', '6.10.3', '6.10.4-rc1', '6.10.4',
+    '6.12.1-rc1', '6.12.1', '6.12.2-rc1', '6.12.2', '6.12.3-rc1', '6.12.3',
+    '7.0.1-rc1', '7.0.1-rc2', '7.0.1', '7.0.2-rc1', '7.0.2-rc2', '7.0.2',
+    '7.0.3', '7.0.4-rc1', '7.0.4',
+  ]  # yapf: disable
+
+  INT_PATTERN = re.compile(r'^[1-9][0-9]*$')
```

Comment:
This regex won't match `0`

Our `semver_index.parse()` will accept and trim numbers with leading 0's, so you can get away with just using the `isdigit()` method on the integer strings.

Reply:
of course, thank you for catching that.